expr/func_movingwindow.go

```diff
@@ -0,0 +1,167 @@
+	if err != nil {
+		return nil, err
+	}
+
+	aggFunc := consolidation.GetAggFunc(consolidation.FromConsolidateBy(s.fn))
+	aggFuncName := fmt.Sprintf("moving%s", strings.Title(s.fn))
+	formatStr := fmt.Sprintf("%s(%%s,\"%s\")", aggFuncName, s.windowSize)
+
+	newName := func(oldName string) string {
+		return fmt.Sprintf(formatStr, oldName)
+	}
+
+	outputs := make([]models.Series, 0, len(series))
+	for _, serie := range series {
+
+		from, points, err := aggregateOnWindow(serie, aggFunc, s.shiftOffset, s.xFilesFactor)
+		if err != nil {
+			return nil, err
+		}
+
+		serie.Target = newName(serie.Target)
+		serie.QueryPatt = newName(serie.QueryPatt)
+		serie.Tags = serie.CopyTagsWith(aggFuncName, s.windowSize)
+		serie.QueryFrom = from
+		serie.Datapoints = points
+
+		outputs = append(outputs, serie)
+	}
+
+	dataMap.Add(Req{}, outputs...)
+	return outputs, nil
+}
+
+func aggregateOnWindow(serie models.Series, aggFunc batch.AggFunc, interval uint32, xFilesFactor float64) (uint32, []schema.Point, error) {
+	out := pointSlicePool.Get().([]schema.Point)
+
+	numPoints := len(serie.Datapoints)
+	serieStart, serieEnd := serie.QueryFrom, serie.QueryTo
+	queryStart := uint32(serieStart + interval)
+
+	if queryStart < serieStart || queryStart > serieEnd {
+		return 0, nil, errors.NewInternalf("query start %d doesn't lie within serie's intervals [%d,%d] ", queryStart, serieStart, serieEnd)
+	}
+
+	// if queryStart is in between points get the index of the point after
+	// else get the index of the point it lies on
+	ptIdx := 0
+	for ptIdx < numPoints && serie.Datapoints[ptIdx].Ts < queryStart {
+		ptIdx++
+	}
+
+	stIdx := 0
+	stTs := serie.Datapoints[stIdx].Ts
+	for ptTs := queryStart; ptTs <= serieEnd && ptIdx < numPoints; ptIdx++ {
+		ptTs = serie.Datapoints[ptIdx].Ts
+
+		// increment start index only if the difference between current and start times
+		// becomes greater than the windowSize (interval)
+		for stIdx <= ptIdx && ptTs-stTs > interval {
+			stIdx++
+			stTs = serie.Datapoints[stIdx].Ts
+		}
+
+		points := serie.Datapoints[stIdx:ptIdx]
+		aggPoint := schema.Point{Val: math.NaN(), Ts: ptTs}
+
+		if pointsXffCheck(points, xFilesFactor) {
+			aggPoint.Val = aggFunc(points)
+		}
+
+		out = append(out, aggPoint)
+	}
+
+	return queryStart, out, nil
+}
+
+func (s *FuncMovingWindow) getWindowSeconds() (uint32, error) {
+
+	// Discard the sign, if present, since we need to consider
+	// preceding datapoints for each point on the graph
+	durStr := s.windowSize
+	if durStr[0] == '-' || durStr[0] == '+' {
+		durStr = durStr[1:]
+	}
+
+	// Note: s.windowSize is not overwritten and will be appended as is to the
+	// new Target, QueryPatt and Tags (keeping it consistent with graphite)
+	// This implies a window size of '-2min' or '+2min' will be captured as is.
+	// However the time operation is always a look back over preceding points.
+	// If this needs to be corrected in the native implementation
+	// uncomment the line below.
+	// s.windowSize = durStr
```

Comment:
no need to account for very unlikely future changes

Reply:
ok. I'll take this out